### PR TITLE
[36] Automatically enter code when aws sso login

### PR DESCRIPTION
**Pull request by Franr**
## What?
Autocomplete the SSO code on the login flow.

## Why?
To improve the user experience.

## References
closes [#36](https://github.com/binbashar/le-docker-leverage-toolbox/issues/36)

## Tests

1. Run `leverage aws sso login`
2. The user code should be set on the browser automatically.

### CODE REVIEW

File: leverage/container.py [hunk 1]
```diff
@@ -422,6 +434,47 @@ def system_exec(self, command):
...
+            if "Then enter the code:" in logs:
+                raw_logger.info(logs)  # container logs already come formatted, so we need to print them raw
+                return logs.split("Then enter the code:")[1].split("\n")[2]
+
+            sleep(self.AWS_SSO_CODE_WAIT_SECONDS)
+
+        raise ExitError(1, "Get SSO code timed-out")
+
+    def get_sso_region(self):
+        # TODO: what about using the .region property we have now? that takes the value from the path of the layer
```
Comment (Franr) [inline leverage/container.py:454]: I know you guys were having a conversation about regions, so would appreciate your feedback here
@angelofenoglio @juanmatias

File: leverage/logger.py [hunk 2]
```diff
@@ -156,3 +157,17 @@ def get_tasks_logger():
     logger.handlers[0].setFormatter(formatter)
 
     return logger
+
+
+def raw_logger():
```
Comment (Franr) [inline leverage/logger.py:162]: any suggestion on how to do this better are welcome!

File: leverage/container.py [hunk 3]
```diff
@@ -379,10 +385,16 @@ class AWSCLIContainer(LeverageContainer):
...
     AWS_CLI_BINARY = "/usr/local/bin/aws"
 
+    # SSO scripts
     AWS_SSO_LOGIN_SCRIPT = "/root/scripts/aws-sso/aws-sso-login.sh"
     AWS_SSO_LOGOUT_SCRIPT = "/root/scripts/aws-sso/aws-sso-logout.sh"
     AWS_SSO_CONFIGURE_SCRIPT = "/root/scripts/aws-sso/aws-sso-configure.sh"
 
+    # SSO constants
+    AWS_SSO_LOGIN_URL = "https://device.sso.{region}.amazonaws.com/?user_code={user_code}"
+    AWS_SSO_CODE_WAIT_SECONDS = 1
```
Comment (diego-ojeda-binbash) [inline leverage/container.py:395]:
```suggestion
    AWS_SSO_CODE_WAIT_SECONDS = 2
```
I'd go with 2 here as I think the feedback message is enough to calm the user's anxiety after running the 'configure sso' command.

File: leverage/container.py [hunk 4]
```diff
@@ -426,6 +438,47 @@ def system_exec(self, command):
...
         return exit_code, output
 
+    def get_sso_code(self, container) -> str:
+        """
+        Find and return the SSO user code by periodically checking the logs.
+        Up until N attempts.
+        """
+        for _ in range(self.AWS_SSO_CODE_ATTEMPTS):
+            # pull logs periodically until we find our SSO code
+            logger.info("Fetching SSO code...")
```
Comment (angelofenoglio) [inline leverage/container.py:448]: I think maybe printing this just once could be enough... buy we can leave this improvement for a future version.

**Review by angelofenoglio — APPROVED** [review]

**Review by diego-ojeda-binbash — APPROVED** [review]

File: leverage/container.py (hunk 4 above)
Comment (Franr) [reply leverage/container.py:448]: future version will be!

File: leverage/logger.py (hunk 2 above)
Comment (Franr) [reply leverage/logger.py:168]: none! going this way

File: leverage/container.py (hunk 1 above)
Comment (Franr) [reply leverage/container.py:458]: right! we will revisit this command then when the conversation comes up

**Pull request merged**